# Exploration Notebook
The purpose of this notebook is to explore the html structure of the webpage listings so as to develop a script for scraping.
## Objectives
<ol>
<li>Establish whether pages are consistent/structured enough for text scraping vs screenshots</li>
<li>Ascertain which tags contain the relevant web page features (amenities section, room listings, price containers etc)</li>
<li></li>
</ol>

# Exploring All Listings Page

In [13]:
from bs4 import BeautifulSoup
import os
import sys
from pathlib import Path
import sys

project_root = str(Path().resolve().parent)
sys.path.insert(0, project_root)

In [11]:
from config import RAW_DIR


In [14]:
with open(RAW_DIR / 'listings_chania.html', encoding='utf-8') as f:
    soup = BeautifulSoup(f, 'html.parser')






In [15]:
listing_links = soup.find_all(class_="bd77474a8e")

In [16]:
len(listing_links)

438

In [17]:
listing_link_urls = []

for listing_link in listing_links:
    listing_link_urls.append(listing_link.get('href'))

listing_link_urls[-10:]





['https://www.booking.com/hotel/gr/villa-avra-crete.en-gb.html?label=gen173bo-10CAsoXEIdc3RhdnJvcy12aWxsYXMtYW1wLWFwYXJ0bWVudHNICVgDaE2IAQGYATO4ARfIAQzYAQPoAQH4AQGIAgGYAgKoAgG4AoiNvM0GwAIB0gIkZGIzNGE3YzgtZWI3YS00N2NkLWI4ODktNDE0YmZmZGJhMGVl2AIB4AIB&aid=304142&ucfs=1&arphpl=1&checkin=2026-04-03&checkout=2026-04-04&dest_id=2006&dest_type=region&group_adults=1&req_adults=1&no_rooms=1&group_children=0&req_children=0&hpos=7&hapos=429&sr_order=popularity&srpvid=dc574f1827e7053c&srepoch=1773141615&all_sr_blocks=128217801_89534957_5_0_0&highlighted_blocks=128217801_89534957_5_0_0&matching_block_id=128217801_89534957_5_0_0&sr_pri_blocks=128217801_89534957_5_0_0__19946&from=searchresults',
 'https://www.booking.com/hotel/gr/villa-amvrosia-crete.en-gb.html?label=gen173bo-10CAsoXEIdc3RhdnJvcy12aWxsYXMtYW1wLWFwYXJ0bWVudHNICVgDaE2IAQGYATO4ARfIAQzYAQPoAQH4AQGIAgGYAgKoAgG4AoiNvM0GwAIB0gIkZGIzNGE3YzgtZWI3YS00N2NkLWI4ODktNDE0YmZmZGJhMGVl2AIB4AIB&aid=304142&ucfs=1&arphpl=1&checkin=2026-04-03&checkout=202

### Findings: 
**Locating Listings:** Based on the above, it appears the same class_ component is used to contain all listing URLs. Need to check if this changes over time and whether it changes between sessions -
Update as of 13/04/2026: They don't seem to change between sessions/over time

**Changing Listing Reservation Dates:** Different parameters in the URL allows for the date to be changed without on-page javascript interaction:

> **URL:**
> https://www.booking.com/hotel/gr/villa-avra-crete.en-gb.html?label=gen173bo-10CAsoXEIdc3RhdnJvcy12aWxsYXMtYW1wLWFwYXJ0bWVudHNICVgDaE2IAQGYATO4ARfIAQzYAQPoAQH4AQGIAgGYAgKoAgG4AoiNvM0GwAIB0gIkZGIzNGE3YzgtZWI3YS00N2NkLWI4ODktNDE0YmZmZGJhMGVl2AIB4AIB&aid=304142&ucfs=1&arphpl=1&checkin=2026-04-08&checkout=2026-04-15&dest_id=2006&dest_type=region&group_adults=1&req_adults=1&no_rooms=1&group_children=0&req_children=0&hpos=7&hapos=429&sr_order=popularity&srpvid=dc574f1827e7053c&srepoch=1773141615&all_sr_blocks=128217801_89534957_5_0_0&highlighted_blocks=128217801_89534957_5_0_0&matching_block_id=128217801_89534957_5_0_0&sr_pri_blocks=128217801_89534957_5_0_0__19946&from=searchresults
> **URL Parameters:**
> {
"label": "gen173bo-10CAsoXEIdc3RhdnJvcy12aWxsYXMtYW1wLWFwYXJ0bWVudHNICVgDaE2IAQGYATO4ARfIAQzYAQPoAQH4AQGIAgGYAgKoAgG4AoiNvM0GwAIB0gIkZGIzNGE3YzgtZWI3YS00N2NkLWI4ODktNDE0YmZmZGJhMGVl2AIB4AIB",
"aid": "304142",
"ucfs": "1",
"arphpl": "1",
"checkin": "2026-04-08",
"checkout": "2026-04-15",
  "dest_id": "2006",
  "dest_type": "region",
  "group_adults": "1",
  "req_adults": "1",
  "no_rooms": "1",
  "group_children": "0",
  "req_children": "0",
  "hpos": "7",
  "hapos": "429",
  "sr_order": "popularity",
  "srpvid": "dc574f1827e7053c",
  "srepoch": "1773141615",
  "all_sr_blocks": "128217801_89534957_5_0_0",
  "highlighted_blocks": "128217801_89534957_5_0_0",
  "matching_block_id": "128217801_89534957_5_0_0",
  "sr_pri_blocks": "128217801_89534957_5_0_0__19946",
  "from": "searchresults"
}

# Property Listing Page Analysis

In [50]:
with open(RAW_DIR / 'elia_palatino_listing_page.html', encoding='utf-8') as f:
    soup = BeautifulSoup(f, 'html.parser')


In [51]:
# Testing f6b6d2a959 to see if it contains all facility elements
facilities = soup.find_all(class_="f6b6d2a959")

In [52]:
len(facilities)

for i, f in enumerate(facilities):
    print(f.text)
    if i > 10:
        break

Free WiFi
Airport shuttle
Non-smoking rooms
Terrace
Superb breakfast
See all 66 facilities
Free WiFi
Airport shuttle
Non-smoking rooms
Terrace
Superb breakfast
Private bathroom


"f6b6d2a959" contains all the facility elements, now I need to extract the text carefully, because sometimes contains multiple text boxes e.g. [**'Cooking class'**, 'Additional charge', 'Off-site']


In [53]:
facility_boxes = soup.find_all(class_="b0bf4dc58f b2f588b43c b14a6d9e99 ee817a4ef2")

In [54]:
list(facility_boxes[31].stripped_strings)

['Cooking class', 'Additional charge', 'Off-site']

# Room type boxes

In [55]:
soup.find?

Signature:
soup.find(
    name: '_FindMethodName' = None,
    attrs: 'Optional[_StrainableAttributes]' = None,
    recursive: 'bool' = True,
    string: 'Optional[_StrainableString]' = None,
    **kwargs: '_StrainableAttribute',
) -> 'Union[_AtMostOneTag, _AtMostOneNavigableString, _AtMostOneElement]'
Docstring:
Look in the children of this PageElement and find the first
PageElement that matches the given criteria.

All find_* methods take a common set of arguments. See the online
documentation for detailed explanations.

:param name: A filter on tag name.
:param attrs: Additional filters on attribute values.
:param recursive: If this is True, find() will perform a
    recursive search of this Tag's children. Otherwise,
    only the direct children will be considered.
:param string: A filter on the `Tag.string` attribute.
:kwargs: Additional filters on attribute values.
File:      c:\users\gabri\documents\projects\tourism_pricing_analytics\.venv\lib\site-packages\bs4\element.py
Type:  

In [56]:
import re

test = soup.find_all(href=re.compile(r"^#RD"))
test

[<a class="hprt-roomtype-link" data-et-click="" data-room-id="217097709" data-room-name="" href="#RD217097709" id="room_type_id_217097709">
 <span class="hprt-roomtype-icon-link">
 Classic Room
 </span>
 </a>,
 <a class="hprt-roomtype-link" data-et-click="" data-room-id="217097702" data-room-name="" href="#RD217097702" id="room_type_id_217097702">
 <span class="hprt-roomtype-icon-link">
 Superior Room
 </span>
 </a>,
 <a class="hprt-roomtype-link" data-et-click="" data-room-id="217097704" data-room-name="" href="#RD217097704" id="room_type_id_217097704">
 <span class="hprt-roomtype-icon-link">
 Deluxe room with sea view and balcony
 </span>
 </a>]

In [57]:
test[0]

<a class="hprt-roomtype-link" data-et-click="" data-room-id="217097709" data-room-name="" href="#RD217097709" id="room_type_id_217097709">
<span class="hprt-roomtype-icon-link">
Classic Room
</span>
</a>

### Observations
- Had to search for a stable string inside html to identify room type boxes:
    - When it start "#RD", seems to be most stable - tried different listings with a few days inbetween and different session/cookies